src/hyperlight_common/src/flatbuffer_wrappers/util.rs

```diff
@@ -169,3 +170,350 @@ impl FlatbufferSerializable for bool {
         }
     }
 }
+
+/// Estimates the required buffer capacity for encoding a FunctionCall with the given parameters.
+/// This helps avoid reallocation during FlatBuffer encoding when passing large slices and strings.
+///
+/// The function aims to be lightweight and fast and run in O(1) as long as the number of parameters is limited
+/// (which it is since hyperlight only currently supports up to 12).
+///
+/// Note: This estimates the capacity needed for the inner vec inside a FlatBufferBuilder. It does not
+/// necessarily match the size of the final encoded buffer. The estimation always rounds up to the
+/// nearest power of two to match FlatBufferBuilder's allocation strategy.
+///
+/// The estimations are numbers used are empirically derived based on the tests below and vaguely based
+/// on https://flatbuffers.dev/internals/ and https://github.com/dvidelabs/flatcc/blob/f064cefb2034d1e7407407ce32a6085c322212a7/doc/binary-format.md#flatbuffers-binary-format
+#[inline] // allow cross-crate inlining (for hyperlight-host calls)
+pub fn estimate_flatbuffer_capacity(function_name: &str, args: &[ParameterValue]) -> usize {
+    let mut estimated_capacity = 20;
+
+    // Function name overhead
+    estimated_capacity += function_name.len() + 12;
+
+    // Parameters vector overhead
+    estimated_capacity += 12 + args.len() * 6;
+
+    // Per-parameter overhead
+    for arg in args {
+        estimated_capacity += 16; // Base parameter structure
+        estimated_capacity += match arg {
+            ParameterValue::String(s) => s.len() + 20,
+            ParameterValue::VecBytes(v) => v.len() + 20,
+            ParameterValue::Int(_) | ParameterValue::UInt(_) => 16,
+            ParameterValue::Long(_) | ParameterValue::ULong(_) => 20,
+            ParameterValue::Float(_) => 16,
+            ParameterValue::Double(_) => 20,
+            ParameterValue::Bool(_) => 12,
```

Comment:
Maybe consider making these values consts to avoid diff magic numbers here and there?

Reply:
I think this is fine IMO, shouldn't need tweaking